src/ray/core_worker/pubsub/pubsub_coordinator.cc

```diff
@@ -0,0 +1,199 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/pubsub/pubsub_coordinator.h"
+
+namespace ray {
+
+void SubscriptionIndex::AddEntry(const ObjectID &object_id, const NodeID &subscriber_id) {
+  auto &subscribing_objects = subscribers_to_objects_[subscriber_id];
+  RAY_CHECK(subscribing_objects.emplace(object_id).second);
+  auto &subscriber_map = objects_to_subscribers_[object_id];
+  RAY_CHECK(subscriber_map.emplace(subscriber_id).second);
+}
+
+const absl::flat_hash_set<NodeID> &SubscriptionIndex::GetSubscriberIdsByObjectId(
+    const ObjectID &object_id) {
+  auto it = objects_to_subscribers_.find(object_id);
+  RAY_CHECK(it != objects_to_subscribers_.end());
+  return it->second;
+}
+
+bool SubscriptionIndex::IsObjectIdExist(const ObjectID &object_id) const {
+  return objects_to_subscribers_.count(object_id) > 0;
+}
+
+bool SubscriptionIndex::IsSubscriberExist(const NodeID &subscriber_id) const {
+  return subscribers_to_objects_.count(subscriber_id) > 0;
+}
+
+int SubscriptionIndex::EraseSubscriber(const NodeID &subscriber_id) {
+  auto subscribing_objects_it = subscribers_to_objects_.find(subscriber_id);
+  if (subscribing_objects_it == subscribers_to_objects_.end()) {
+    return 0;
+  }
+
+  auto &subscribing_objects = subscribing_objects_it->second;
+  for (auto object_id_it = subscribing_objects.begin();
+       object_id_it != subscribing_objects.end();) {
+    auto current = object_id_it++;
+    const auto &object_id = *current;
+    // Erase the subscriber from the object map.
+    auto subscribers_it = objects_to_subscribers_.find(object_id);
+    if (subscribers_it == objects_to_subscribers_.end()) {
+      continue;
+    }
+    auto &subscribers = subscribers_it->second;
+    subscribers.erase(subscriber_id);
+    if (subscribers.size() == 0) {
+      objects_to_subscribers_.erase(subscribers_it);
+    }
+  }
```

Comment:
Nit: Why not use a range-based for loop?
```suggestion
  for (const auto &object_id : subscribing_objects) {
    // Erase the subscriber from the object map.
    auto subscribers_it = objects_to_subscribers_.find(object_id);
    if (subscribers_it == objects_to_subscribers_.end()) {
      continue;
    }
    auto &subscribers = subscribers_it->second;
    subscribers.erase(subscriber_id);
    if (subscribers.size() == 0) {
      objects_to_subscribers_.erase(subscribers_it);
    }
  }
```

Reply:
Ah good catch. I originally had an implementation that needs to delete the array itself within a loop and didn't fix it when I changed that logic.